pkg/runtime/wfengine/README.md

```diff
@@ -64,6 +64,36 @@ In the future, instructions will be added for how to tests in a more automated w
 
-TODO: More information on activity actors, the details of state storage, and information about how reminders are used to ensure reliability.
\ No newline at end of file
+### State storage
+
+Each workflow actor saves its state using the following keys:
+
+* `metadata`: Contains meta information about the workflow as a JSON blob. Includes information such as the length of the inbox, the length of the history, and a UUID representing the workflow generation (for cases where the instance ID gets reused). The length information is used to determine which keys need to be read or written to when loading or saving workflow state updates.
+* `inbox-NNNNNN`: Multiple keys containing an ordered list of workflow inbox events. Each key holds the data for a single event. The inbox is effectively a FIFO queue of events that the workflow needs to process, with items removed from the earlier indeces and added to the end indices.
+* `history-NNNNNN`: Multiple keys containing an ordered list of history events. Each key holds the data for a single event. History events are only added and never removed, except in the case of "continue as new", where all history events are purged.
+* `customStatus`: Contains a user-defined workflow status value.
+
+The `inbox-NNNNNN` and `history-NNNNNN` key schemes are used to enable arbitrarily large amounts of data. These schemes are also designed for efficient updates. An alternate design would be to store the workflow history as a blob in a single key. However, this would limit the maximum size of the history and would make updates more expensive, since the full history would need to be serialized instead of just inserting incremental additions (the history is an append-only log of events).
+
+The tradeoff with this key scheme design is that loading workflow state becomes more expensive since it's spread out across multiple keys. This is mitigated by the fact that actor state can be cached in memory, removing the need for any reads while the actors are active. However, it could be a problem if workflow histories get large and if actors get moved around or activated frequently.
+
+Below is an example of what keys would be used to store the state of a simple workflow execution with ID '797f67f0c10846f592d0ac82dea1f248', as shown using `redis-cli`.
+
+```
+127.0.0.1:6379> keys *797f67f0c10846f592d0ac82dea1f248*
+1) "myapp||dapr.internal.wfengine.workflow||797f67f0c10846f592d0ac82dea1f248||history-000002"
```

Comment:
My only concern with this is that there is a limitation on the number of keys that a single redis instance can deal with - this limit may far exceed the practical ceiling we would see but if we don't purge the data we could end up hitting it. 

Thoughts?

Reply:
I'm thinking we cap the maximum history size to 10K and the maximum inbox size to 1K. This is something we probably want to do anyways to limit the impact of runaway workflows. Also, good point about data purging. I'm going to add a work item to revisit both of these in a separate data governance PR.